### PR TITLE
introduce expr/constant into fdSet with uniqueID and refine some op's logic

### PR DESCRIPTION
**Pull request by AilinKid**
Signed-off-by: ailinkid <314806019@qq.com>

### What problem does this PR solve? 
1. introduce expr/constant into fdSet with uniqueID allocated
2. extract constant/equivalence/notnull attribute from select condition
3. refine some logic about groupby/projection/datasource
4. add some test about extractFD's behavior

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/util.go [hunk 1]
```diff
@@ -166,6 +166,161 @@ func extractColumns(result []*Column, expr Expression, filter func(*Column) bool
 	return result
 }
 
+func ExtractEquivalenceColumns(result [][]interface{}, exprs []Expression) [][]interface{} {
```
Comment (winoros) [inline expression/util.go]: Add header comment

File: expression/util.go [hunk 2]
```diff
@@ -166,6 +166,161 @@ func extractColumns(result []*Column, expr Expression, filter func(*Column) bool
...
+	for _, expr := range exprs {
+		result = extractEquivalenceColumns(result, expr)
+	}
+	return result
+}
+
+func extractEquivalenceColumns(result [][]interface{}, expr Expression) [][]interface{} {
```
Comment (winoros) [inline expression/util.go]: Why not return `expression.Expression`?

File: planner/core/optimizer.go [hunk 3]
```diff
@@ -375,6 +375,11 @@ func enableParallelApply(sctx sessionctx.Context, plan PhysicalPlan) PhysicalPla
 	return plan
 }
 
+// LogicalOptimize is just exported for test.
+func LogicalOptimize(ctx context.Context, flag uint64, logic LogicalPlan) (LogicalPlan, error) {
```
Comment (winoros) [inline planner/core/optimizer.go]:
```suggestion
func LogicalOptimizeForTest(ctx context.Context, flag uint64, logic LogicalPlan) (LogicalPlan, error) {
```

File: expression/util.go (hunk 1 above)
Comment (winoros) [inline expression/util.go]: Actually, the equiv class can be maintained via the selection/join?

File: expression/util.go (hunk 2 above)
Comment (AilinKid) [reply expression/util.go]: Nice catch, make sense

File: expression/util.go (hunk 1 above)
Comment (AilinKid) [reply expression/util.go]: Yes, and what ideas about `be maintained via the selection/join` ?

Comment (AilinKid) [reply expression/util.go]: seem selection's condition will change dynamically for PropagateConstant/Decorrelate and so on, maintaining work should catch up in these position. i'll leave it here for future work.

**Review by winoros — APPROVED** [review]

**Pull request merged**